File: tests/api/typescript/src/state.ts
export interface Chunk {
  offset: number;
  data: string;
  checksum: string;
  checksumPrevious: string | null;
  state: "partial" | "complete";
  cursor: string | null;
}

export interface OperationInstance {
  requestId: string;
  op: string;
  state: "accepted" | "pending" | "complete" | "error";
  result?: unknown;
  error?: { code: string; message: string };
  retryAfterMs: number;
  createdAt: number;
  expiresAt: number;
  chunks?: Chunk[];
}

let operationInstances = new Map<string, OperationInstance>();

export function createInstance(requestId: string, op: string): OperationInstance {
  const instance: OperationInstance = {
    requestId,
    op,
    state: "accepted",
    retryAfterMs: 100,
    createdAt: Math.floor(Date.now() / 1000),
    expiresAt: Math.floor(Date.now() / 1000) + 3600,
  };
  operationInstances.set(requestId, instance);
  return instance;
}

export function transitionTo(
  requestId: string,
  state: OperationInstance["state"],
  data?: { result?: unknown; error?: { code: string; message: string }; chunks?: Chunk[] },
): OperationInstance | null {
  const instance = operationInstances.get(requestId);
  if (!instance) return null;
  instance.state = state;
  if (data?.result !== undefined) instance.result = data.result;
  if (data?.error) instance.error = data.error;
  if (data?.chunks) instance.chunks = data.chunks;
  return instance;
}

export function getInstance(requestId: string): OperationInstance | null {
  return operationInstances.get(requestId) || null;
}

export function resetInstances(): void {
  operationInstances = new Map();
}

export function computeSha256(data: string): string {
  const hash = new Bun.CryptoHasher("sha256");
  hash.update(data);
  return `sha256:${hash.digest("hex")}`;
}

export function buildChunks(data: string, chunkSize: number = 512): Chunk[] {
  const chunks: Chunk[] = [];
  let offset = 0;
  let previousChecksum: string | null = null;

  while (offset < data.length) {
    const end = Math.min(offset + chunkSize, data.length);
    const chunkData = data.slice(offset, end);
    const checksum = computeSha256(chunkData);
    const isLast = end >= data.length;
    const cursor = isLast ? null : btoa(String(end));

    chunks.push({
      offset,
      data: chunkData,
      checksum,
      checksumPrevious: previousChecksum,
      state: isLast ? "complete" : "partial",
      cursor,
    });

    previousChecksum = checksum;
    offset = end;
  }

  return chunks;
}

File: tests/api/go/state.go
package main

import (
	"crypto/sha256"
	"encoding/base64"
	"fmt"
	"sync"
	"time"
)

type chunk struct {
	Offset           int     `json:"offset"`
	Data             string  `json:"data"`
	Checksum         string  `json:"checksum"`
	ChecksumPrevious *string `json:"checksumPrevious"`
	State            string  `json:"state"`
	Cursor           *string `json:"cursor"`
}

type operationInstance struct {
	RequestID    string
	Op           string
	State        string // "accepted", "pending", "complete", "error"
	Result       interface{}
	Error        map[string]interface{}
	RetryAfterMs int
	CreatedAt    int64
	ExpiresAt    int64
	Chunks       []chunk
}

var (
	instances   = make(map[string]*operationInstance)
	instancesMu sync.RWMutex
)

func createInstance(requestID string, op string) *operationInstance {
	instance := &operationInstance{
		RequestID:    requestID,
		Op:           op,
		State:        "accepted",
		RetryAfterMs: 100,
		CreatedAt:    time.Now().Unix(),
		ExpiresAt:    time.Now().Add(3600 * time.Second).Unix(),
	}

	instancesMu.Lock()
	instances[requestID] = instance
	instancesMu.Unlock()
	return instance
}

func transitionTo(requestID string, state string, data map[string]interface{}) *operationInstance {
	instancesMu.Lock()
	defer instancesMu.Unlock()

	instance, exists := instances[requestID]
	if !exists {
		return nil
	}
	instance.State = state
	if data != nil {
		if result, ok := data["result"]; ok {
			instance.Result = result
		}
		if errVal, ok := data["error"]; ok {
			if errMap, ok := errVal.(map[string]interface{}); ok {
				instance.Error = errMap
			}
		}
		if chunksVal, ok := data["chunks"]; ok {
			if chunks, ok := chunksVal.([]chunk); ok {
				instance.Chunks = chunks
			}
		}
	}
	return instance
}

func getInstance(requestID string) *operationInstance {
	instancesMu.RLock()
	defer instancesMu.RUnlock()
	return instances[requestID]
}

func resetInstances() {
	instancesMu.Lock()
	defer instancesMu.Unlock()
	instances = make(map[string]*operationInstance)
}

func computeSha256(data string) string {
	h := sha256.Sum256([]byte(data))
	return fmt.Sprintf("sha256:%x", h)
}

func buildChunks(data string, chunkSize int) []chunk {
	if chunkSize <= 0 {
		chunkSize = 512
	}
	var chunks []chunk
	offset := 0
	var previousChecksum *string

	for offset < len(data) {
		end := offset + chunkSize
		if end > len(data) {
			end = len(data)
		}
		chunkData := data[offset:end]
		checksum := computeSha256(chunkData)
		isLast := end >= len(data)

		var cursor *string
		if !isLast {
			c := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%d", end)))
			cursor = &c
		}

		state := "partial"
		if isLast {
			state = "complete"
		}

		chunks = append(chunks, chunk{
			Offset:           offset,
			Data:             chunkData,
			Checksum:         checksum,
			ChecksumPrevious: previousChecksum,
			State:            state,
			Cursor:           cursor,
		})

		prev := checksum
		previousChecksum = &prev
		offset = end
	}

	return chunks
}

func nowISO() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05.000Z")
}
